src/wallet/wallet_tx_builder.cpp

```diff
@@ -0,0 +1,683 @@
+            }
+        }, selector.GetPattern());
+        assert(fvk.has_value());
+
+        internalOVK = fvk.value().ToInternalOutgoingViewingKey();
+        externalOVK = fvk.value().ToExternalOutgoingViewingKey();
+    } else if (!spendable.saplingNoteEntries.empty()) {
+        std::optional<SaplingDiversifiableFullViewingKey> dfvk;
+        std::visit(match {
+            [&](const libzcash::SaplingPaymentAddress& addr) {
+                libzcash::SaplingExtendedSpendingKey extsk;
+                assert(pwalletMain->GetSaplingExtendedSpendingKey(addr, extsk));
+                dfvk = extsk.ToXFVK();
+            },
+            [&](const UnifiedAddress& ua) {
+                auto ufvk = pwalletMain->GetUFVKForAddress(ua);
+                // This is safe because spending key checks will have ensured that we
+                // have a UFVK corresponding to this address, and Sapling notes will
+                // not have been selected if the UFVK does not contain a Sapling key.
+                dfvk = ufvk.value().GetSaplingKey().value();
+            },
+            [&](const UnifiedFullViewingKey& ufvk) {
+                // Sapling notes will not have been selected if the UFVK does not contain
+                // a Sapling key.
+                dfvk = ufvk.GetSaplingKey().value();
+            },
+            [&](const AccountZTXOPattern& acct) {
+                // By definition, we have a UFVK for every known account.
+                auto ufvk = pwalletMain->GetUnifiedFullViewingKeyByAccount(acct.GetAccountId());
+                // Sapling notes will not have been selected if the UFVK does not contain
+                // a Sapling key.
+                dfvk = ufvk.value().GetSaplingKey().value();
+            },
+            [&](const auto& other) {
+                throw std::runtime_error("SelectOVKs: Selector cannot select Sapling notes.");
+            }
+        }, selector.GetPattern());
+        assert(dfvk.has_value());
+
+        auto ovks = dfvk.value().GetOVKs();
+        internalOVK = ovks.first;
+        externalOVK = ovks.second;
+    } else if (!spendable.utxos.empty()) {
+        std::optional<transparent::AccountPubKey> tfvk;
+        std::visit(match {
+            [&](const CKeyID& keyId) {
+                tfvk = pwalletMain->GetLegacyAccountKey().ToAccountPubKey();
+            },
+            [&](const CScriptID& keyId) {
+                tfvk = pwalletMain->GetLegacyAccountKey().ToAccountPubKey();
+            },
+            [&](const UnifiedAddress& ua) {
+                // This is safe because spending key checks will have ensured that we
+                // have a UFVK corresponding to this address, and transparent UTXOs will
+                // not have been selected if the UFVK does not contain a transparent key.
+                auto ufvk = pwalletMain->GetUFVKForAddress(ua);
+                tfvk = ufvk.value().GetTransparentKey().value();
```

Comment:
```suggestion
                tfvk = ufvk.value().GetTransparentKey();
```
Similar argument to `fvk` and `dfvk` above.